<a href="https://colab.research.google.com/github/sofiacorreao/Analitica/blob/main/Reto1_Sofia_Correa_O.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np #Funciones matematicas
import pandas as pd #Comunicación con Excel
from sklearn.naive_bayes import GaussianNB #Funcion Gaussiana
from sklearn.metrics import confusion_matrix #La metrica de medicion

In [ ]:
nxl="/content/1. BD1_Longitudinal Survey.xlsx"
XDB=pd.read_excel(nxl,sheet_name=0).dropna() #Traigo los datos
 #Eliminar datos faltantes, celdas vacias

#Seleccionamos las variables de interes
XDB=XDB[['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize','Urban']]
XDB

#Datos de Entrada - np.array ()
XD=XDB[['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize']]
#Datos de salida - Valor deseado
yd=np.array(XDB['Urban'] )
print (XDB)

       Age  Mother_Edu  Father_Edu  Height  Weight   Income  Self_Esteem  \
1       20         5.0         8.0    62.0   120.0      0.0         16.0   
5       19        12.0        12.0    64.0   200.0  40000.0         27.0   
6       15        12.0        12.0    65.0   131.0  25000.0         26.0   
7       21         9.0         6.0    65.0   179.0  27400.0         23.0   
8       16        12.0        10.0    66.0   145.0  52000.0         26.0   
...    ...         ...         ...     ...     ...      ...          ...   
12557   20        12.0        12.0    60.0   110.0   1000.0         20.0   
12565   20        12.0        11.0    63.0   152.0  38000.0         24.0   
12647   21        16.0        12.0    70.0   180.0  43600.0         30.0   
12651   21        16.0        16.0    73.0   170.0  40000.0         27.0   
12678   21         8.0         8.0    70.0   180.0  17000.0         19.0   

       FamilySize  Urban  
1               5    1.0  
5               4    1.0  
6     

In [ ]:
mnb=GaussianNB() #Llamamos el modelo
mnb.fit(XD,yd)
ydp=mnb.predict(XD) #Muestra los datos de salida

#Parametros del modelo por categoria
u=mnb.theta_ #Se estiman las medias
sigma=np.sqrt(mnb.var_) #Se estiman las varianzas

#Las variaciones de las variables
LI=u-sigma
LS=u+sigma

print('La media de las variables por categoria:')
print (u)
print('El limite inferior de la variación de las variables por categoria es:')
print(LI)
print('El limite superior de la variación de las variables por categoria es:')
print (LS)

#Métrica de evaluación
cm=confusion_matrix(yd,ydp)

La media de las variables por categoria:
[[1.83810742e+01 1.07612958e+01 1.03324808e+01 6.72148338e+01
  1.48857630e+02 3.03672506e+04 2.20272805e+01 4.74168798e+00]
 [1.83569845e+01 1.11769401e+01 1.12565410e+01 6.69031042e+01
  1.44760976e+02 3.34278929e+04 2.26004435e+01 4.68780488e+00]]
El limite inferior de la variación de las variables por categoria es:
[[ 1.58996876e+01  7.58892579e+00  6.46311334e+00  6.31369799e+01
   1.18746964e+02 -1.33465823e+03  1.78344685e+01  2.16621834e+00]
 [ 1.58811112e+01  7.79013864e+00  7.15097738e+00  6.26897933e+01
   1.14669654e+02 -2.69533531e+03  1.84154526e+01  2.24996381e+00]]
El limite superior de la variación de las variables por categoria es:
[[2.08624608e+01 1.39336659e+01 1.42018483e+01 7.12926877e+01
  1.78968296e+02 6.20691595e+04 2.62200925e+01 7.31715762e+00]
 [2.08328577e+01 1.45637416e+01 1.53621047e+01 7.11164151e+01
  1.74852297e+02 6.95511211e+04 2.67854344e+01 7.12564594e+00]]


**Determinar el intervalo en el cual se encuentra la variacion de la edad**

R// El limite inferior en urbano es 1.58996876e+01

El limite superior en urbano  es 2.08624608e+01

El limite inferior en rural es  1.58811112e+01

El limite superior en rural es 2.08328577e+01

**Determinar el intervalo en el cual se encuentra la altura**

R// El limite inferior en urbano es 6.31369799e+01

El limite superior en urbano  es 7.12926877e+01

El limite inferior en rural es  6.26897933e+01

El limite superior en rural es  7.11164151e+01

**Determinar el intervalo en el cual se encuentra el peso**

R// El limite inferior en urbano es 1.48857630e+02

El limite superior en urbano  es 1.78968296e+02

El limite inferior en rural es  1.14669654e+02

El limite superior en rural es  1.74852297e+02

In [ ]:
VN=cm[0,0] #Verdaderos negativos
FP=cm[0,1] #Falsos positivos
FN=cm[1,0] #Falsos negativos
VP=cm[1,1] #Verdaderos positivos

Exactitud=(VP+VN)/len(XD)
print('La Exactitud es:',Exactitud) #Comportamiento general

TE=(FN+FP)/len(XD)
print('La Tasa de Error es:',TE) #Porcentaje de Equivocaciones

Sensibilidad=VP/(VP+FP)
print('La Sensibilidad es:',Sensibilidad) #Como se comporto frente a los positivos

Especificidad=VN/(VN+FN)
print('La Especificidad es:',Especificidad) #Como se comportaron frente a los negativos

La Exactitud es: 0.7932430054548654
La Tasa de Error es: 0.2067569945451346
La Sensibilidad es: 0.793729082261758
La Especificidad es: 0.3333333333333333


¿Para una persona con las siguientes características (34,9,6,65,135,50000,30,3)  indicar si vive en la ciudad o las áreas rurales

In [ ]:
Xun=np.array([[34,9,6,65,135,50000,30,3]])
ydun=mnb.predict(Xun)
print(f'Para una persona con las caracteristicas dadas el pronostico es: {ydun}, es decir, vive en la zona urbana')

Para una persona con las caracteristicas dadas el pronostico es: [1.], es decir, vive en la zona urbana


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Cual es el número de datos total de la base de datos?

In [ ]:
total_datos = XDB.shape[0]
print(f'El total de datos que hay en la base de datos es: {total_datos}')

El total de datos que hay en la base de datos es: 5683


In [ ]:
total_datos_urbanos = XDB[XDB['Urban'] == 1].shape[0]
total_datos_urbanos

4510

Qué proporcion del total de los datos corresponde al area rural?

In [ ]:
total_datos_rural = XDB[XDB['Urban'] == 0].shape[0]
total_datos_rural

1173

In [ ]:
tasa_rural = total_datos_rural / total_datos
print(f'La tasa de personas que hacen parte del area rural es de: {tasa_rural}')

La tasa de personas que hacen parte del area rural es de: 0.20640506774590886


Que proporcion del total de los datos corresponde al area urbana?

In [ ]:
tasa_urbanos = total_datos_urbanos / total_datos
print(f'La tasa de personas que hacen parte del area urbana es de: {tasa_urbanos}')

La tasa de personas que hacen parte del area urbana es de: 0.7935949322540912


**Análisis:**

La Exactitud es 0.7932430054548654 lo que indica un rendimiento global razonablemente bueno, pero no excelente.

La Tasa de Error es 0.2067569945451346, complementa la exactitud, mostrando el porcentaje de predicciones incorrectas.

La Sensibilidad es 0.793729082261758, esto indica que el modelo tiene buen desempeño al identificar positivos, lo cual es bueno si los falsos negativos son críticos.

La Especificidad es 0.3333333333333333, este bajo valor indica que el modelo no es muy efectivo para identificar los negativos, lo que puede ser problemático si los falsos positivos son costosos.

En conclusión, el desempeño del modelo en general es bueno ya que tiene una alta exactitud y sensibilidad, sin embargo puede mejorar la especifidad para que la tasa de error disminuya y haya mas efectividad